node/src/components/linear_chain_sync.rs

```diff
@@ -301,127 +313,152 @@ where
+                        // If we do, it's a bug.
+                        assert_eq!(*block.hash(), block_hash, "Block hash mismatch.");
+                        trace!(%block_hash, "linear block found in the local storage.");
+                        self.block_downloaded(rng, effect_builder, block.header())
+                    }
+                    BlockByHashResult::FromPeer(block, peer) => {
+                        self.metrics.observe_get_block_by_hash();
+                        trace!(%block_hash, %peer, "linear chain block downloaded from a peer");
+                        if *block.hash() != block_hash {
+                            warn!(
+                                "block hash mismatch. Expected {} got {} from {}.",
+                                block_hash,
+                                block.hash(),
+                                peer
+                            );
+                            // NOTE: Signal misbehaving validator to networking layer.
+                            self.peers.ban(&peer);
+                            return self.handle_event(
+                                effect_builder,
+                                rng,
+                                Event::GetBlockHashResult(
+                                    block_hash,
+                                    BlockByHashResult::Absent(peer),
+                                ),
+                            );
+                        }
+                        self.peers.success(peer);
+                        self.block_downloaded(rng, effect_builder, block.header())
                     }
-                    self.peers.success(peer);
-                    self.block_downloaded(rng, effect_builder, block.header())
-                }
-            },
-            Event::GetDeploysResult(fetch_result) => match fetch_result {
-                event::DeploysResult::Found(block_header) => {
-                    let block_hash = block_header.hash();
-                    trace!(%block_hash, "deploys for linear chain block found");
-                    // Reset used peers so we can download next block with the full set.
-                    self.peers.reset(rng);
-                    // Execute block
-                    let finalized_block: FinalizedBlock = (*block_header).into();
-                    effect_builder.execute_block(finalized_block).ignore()
                 }
-                event::DeploysResult::NotFound(block_header, peer) => {
-                    let block_hash = block_header.hash();
-                    trace!(%block_hash, %peer, "deploy for linear chain block not found. Trying next peer");
-                    self.peers.failure(&peer);
-                    match self.peers.random() {
-                        None => {
-                            error!(%block_hash,
+            }
+            Event::GetDeploysResult(fetch_result) => {
+                self.metrics.observe_get_deploy();
```

Comment:
```suggestion
                self.metrics.observe_get_deploys();
```

Reply:
Done in https://github.com/CasperLabs/casper-node/pull/786/commits/569e4545d3503d2fd10dc876ba54be4045a74fd2